marshal/admin.go

```diff
@@ -155,35 +155,57 @@ func (a *consumerGroupAdmin) releaseClaims(resetOffset bool) error {
 	}
 }
 
-// heartbeatLoop hearbeats as if we had a claim to this partition and were simply
+// Send a single heartbeat using the given offset.
+func (a *consumerGroupAdmin) heartbeat(
+	topic string, partID int, offset int64) (ok bool) {
+	// If we fail to heartbeat, record this in claimHealth.
+	// The Admin will take care of cleaning up other claims.
+	if err := a.marshaler.Heartbeat(topic, partID, offset); err != nil {
+		log.Errorf("[%s:%d] Admin failed to heartbeat. It is now unhealthy "+
+			"and will not reset offsets.", topic, partID)
+		atomic.StoreInt32(a.claimHealth, 0)
+		return false
+	}
+	return true
+}
+
+// heartbeatLoop heartbeats as if we had a claim to this partition and were simply
 // not reading past where the previous owner had left off.
+//
+// Callers can signal for the heartbeatLoop to stop by sending a message on the
+// stopHeartbeat channel; the heartbeatLoop will close the heartbeatStopped
+// channel after it has in fact stopped.
 func (a *consumerGroupAdmin) heartbeatLoop(
-	topic string, partID int, lastOffset int64, stopChan chan struct{}) {
-
+	topic string, partID int, lastOffset int64,
+	stopHeartbeat, heartbeatStopped chan struct{}) {
+
+	defer close(heartbeatStopped)
+	// We send the first heartbeat right away (instead of waiting for an entire
+	// heartbeat interval to pass).
+	if ok := a.heartbeat(topic, partID, lastOffset); !ok {
+		return
+	}
 	for {
 		select {
 		// Stop claimHealth either when all topic, partitions have been successfully claimed,
 		// or the Admin has failed to do so and needs to abort.
-		case <-stopChan:
+		case <-stopHeartbeat:
 			return
-		default:
-			// If we fail to heartbeat, record this in claimHealth.
-			// The Admin will take care of cleaning up other claims.
-			if err := a.marshaler.Heartbeat(topic, partID, lastOffset); err != nil {
-				log.Errorf("[%s:%d] Admin failed to heartbeat. It is now unhealthy "+
-					"and will not reset offsets.", topic, partID)
-				atomic.StoreInt32(a.claimHealth, 0)
+		// Note that waiting for the heartbeat interval in a select statement (instead of
+		// using time.Sleep) allows the heartbeat to stop right away.
+		case <-time.After(<-a.marshaler.cluster.jitters):
```

Comment:
I don't think this does it, now it will be between 30 and sixty seconds before you successfully heartbeat the first time.

I think what you need is something like this

```
var nextHeartbeatChan <-chan Time
nextHeartbeatChan = make(chan Time)
close(nextHeartbeatChan)

for {
  select {
  case <- stopHeartbeat:
    return

  case <-nextHeartbeatChan:

   < do stuff >
  nextHeartbeatChan = time.After(<-a.marshaller.cluster.jitters)
  }

}
```

This way it will happen immediately the first time, then after the jiter each subsequent time.